src/socket.rs

```diff
@@ -0,0 +1,335 @@
+use crate::arch::Syscalls;
+use crate::close;
+use crate::{result, syscall};
+use core::mem::size_of_val;
+use libc::{
+    sockaddr, socklen_t, AF_INET, AF_INET6, SOCK_CLOEXEC, SOCK_DGRAM, SOCK_NONBLOCK, SOCK_RAW,
+    SOCK_RDM, SOCK_SEQPACKET, SOCK_STREAM,
+};
+use std::net::SocketAddr;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::{
+    io,
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+pub struct Fd(usize);
+
+impl AsRawFd for Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0 as RawFd
+    }
+}
+
+impl AsRawFd for &Fd {
+    fn as_raw_fd(&self) -> RawFd {
+        (*self).as_raw_fd()
+    }
+}
+
+impl Drop for Fd {
+    fn drop(&mut self) {
+        unsafe {
+            close(self).ok();
+        }
+    }
+}
+
+fn socket_addr(addr: SocketAddr) -> (*const sockaddr, socklen_t) {
+    match addr {
+        SocketAddr::V4(ref addr) => (
+            addr as *const _ as *const sockaddr,
+            size_of_val(addr) as socklen_t,
+        ),
+        SocketAddr::V6(ref addr) => (
+            addr as *const _ as *const sockaddr,
+            size_of_val(addr) as socklen_t,
+        ),
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+#[repr(i32)]
+pub enum SockType {
+    Stream = SOCK_STREAM,
+    Datagram = SOCK_DGRAM,
+    SeqPacket = SOCK_SEQPACKET,
+    Raw = SOCK_RAW,
+    Rdm = SOCK_RDM,
+}
+
+#[derive(Clone, Copy)]
+pub struct SockFlags(i32);
+
+impl SockFlags {
+    pub fn new() -> Self {
+        Self(0)
+    }
+
+    pub fn nonblock(mut self) -> Self {
+        self.0 |= SOCK_NONBLOCK;
+        self
+    }
+
+    pub fn cloexec(mut self) -> Self {
+        self.0 |= SOCK_CLOEXEC;
+        self
+    }
+}
+
+#[inline]
+pub fn socket(addr: SocketAddr, sock: SockType, flags: SockFlags) -> io::Result<Fd> {
```

Comment:
I'm not sure if these functions can be marked safe.
I'd rather keep them unsafe until we're pretty damn sure (Open Question 1, https://github.com/elichai/syscalls-rs#open-questions )

Reply:
all inputs and outputs are rust types, so if the function is correct it should be safe no?

Reply:
if it's correct and non of the inputs we're passing can ever have UB in the kernel? then yeah it's safe.
Personally I'm not yet 100% sure if we can guarantee this, but I might be too conservative and catious.

Reply:
I was pleasantly surprised that whenever I made a mistake, the kernel kindly gave me an error message instead of a segfault.

Reply:
I'm with you. but sadly observing something and guaranteeing it are 2 different things

Reply:
I think we should assume the kernel + fn to be bug free when deciding whether to mark a function unsafe. If we mark everything unsafe it looses it's importance. Unsafe means the user needs to pay special attention, not review the code + kernel.